### ISSUE TITLE
Disconnection over time (10 mins), no error?

### ISSUE THREAD

**Issue by lexyoussef** [swiftyspiffy/TwitchLib#161]
I think this could be related to: https://github.com/swiftyspiffy/TwitchLib/issues/148

I'm not quite sure what to do at this point. Running locally (Windows 10 environment), my app runs fine, TwitchClient keeps running as expected for extended periods of time. However, when deploying this as an Azure App Service, the TwitchClient stops functioning (e.g, no longer fires OnMessageReceived) after about 10 minutes. I've enabled logging, I can't seem to catch any errors.

Any information or tips on getting this to function on an Azure App Service would be greatly appreciated! :)

**Comment by swiftyspiffy** [comment]
Hey. Can't help a ton at the moment, but does your Azure instance have Always On set to true? As I understand it, if this setting is set to false, the web app will stop after a certain amount of time.

For this setting to be on, you need to upgrade from the Shared plan I think.

**Comment by lexyoussef** [comment]
It wasn't on before, but I've turned it on and I'm still experiencing the same issue. Does it take time for the change to take place? Is there another setting that should be on?

**Comment by swiftyspiffy** [comment]
It should be instant. That being said, that's the only thing I can think of that would cause the app to close after a few minutes. Perhaps try restarting the instance after setting it to true?

**Comment by lexyoussef** [comment]
Restarted it when I first changed it, but I'm trying a restart again and see if that makes a difference.

**Comment by swiftyspiffy** [comment]
Does the instance stay alive after the chat client disconnects, or is the entire instance dying after a few minutes?

**Comment by lexyoussef** [comment]
Instance is still staying alive, and the controller actions that depend on the client are functioning without error, just getting no messages when I should be.

I seem to have come across an error. I'm looking into it now and will let you know if I make progress soon.

**Comment by lexyoussef** [comment]
Hmm, well I don't want to jinx it but it seems to be running longer than it ever has (granted I've only been testing for about 15 minutes). I had an error of my own that I addressed (which wasn't related to this). I republished the application, and I haven't had issues with the client yet.

I'll give an update either later today or first thing tomorrow morning, but I certainly hope it's addressed.

**Comment by swiftyspiffy** [comment]
Awesome! Good work. Ya, very interested in how this goes. Thanks for the updates!

**Comment by lexyoussef** [comment]
Bad news. It didn't work. :(

Might've just been a fluke that it lasted a few minutes longer.

If no messages are received for a certain amount of time, will that cause the client to stop working? Granted, I've done similar tests running in localhost with no messages being sent for about an hour, and it'll still work after... hmmm

**Comment by swiftyspiffy** [comment]
Damn.

The client sends and receives PING/PONG messages that keep the connection alive. I had a similar problem when attempting to use Azure but figured it was something I was doing wrong. As far as I can tell, it's an issue that's isolated to Azure Web Apps.

For what it's worth, several applications I have deployed on AWS EC2 have an up time of several weeks at the moment.

**Comment by lexyoussef** [comment]
... sigh... and I just got everything set up all nice and pretty on azure hahaha.

I'm going to reach out to Microsoft to see if theres anything else that can be done, and if not, I'll look into migrating over to EC2.

Thanks for your help though! Every bit helps :D

**Comment by swiftyspiffy** [comment]
Ha, ya that sucks. There have been others with your exact problem as you mentioned above. Kind of unfortunate.

Perhaps you might try a different library and see if it works on Azure or if it has similar problems. If it works, then I'd be interested in knowing which library that is.

**Comment by lexyoussef** [comment]
Hmmm okay, I'll do a bit more research... If I find anything I'll post it back here.

Thanks again!

**Comment by swiftyspiffy** [comment]
Ya man, no problem!

**Comment by lexyoussef** [comment]
Alright, so I haven't found an alternative library as of yet. But while poking around I was reading through the official Twitch API documentation, and I was wondering what limitations there are in using the TwitchClient? I saw for PubSub there's a limit of 10 simultaneous connections per ip address, does that also apply to the TwitchClient?

I also noticed the following: "If you send more than 20 commands or messages to the server within 30 seconds, you will be locked out for 30 minutes." 

What would be considered a message, for example? Does that include any events fired from subscribing to something like OnMessageReceived?

I don't necessarily think it's related to the same problem, but it's possible that in my implementation I may be doing something that's causing these disconnections to occur.

Thanks!

**Comment by swiftyspiffy** [comment]
So, every TwitchClient object you create and invoke Connect() on would be considered a connection. Keep in mind a TwitchClient object can join many channels under a single connection. As far as normal chat, I can't recall what the limitations are, but I think they're larger than pubsub.

As far as limitations on sending, any chat command like purging and changing color etc and chat message is considered towards the messages sent total. Subscribing to events does not count (the library handles the events using data that is already normally available).

Receiving messages does not count towards rate limiting, only sending messages/commands.

**Comment by lexyoussef** [comment]
Ok, well I'm not doing anything outside the limits... I'm going to tinker around some more and see if I can get any improvements on Azure, but this may end up being a switch to EC2.

**Comment by lexyoussef** [comment]
Hey again, I just wanted to update you on some of my findings.

I've noticed that when triggering events to occur about once every 1-3 minutes (such as OnMessageReceived), the client appears to work as long as one would expect on Azure.

After about 1 hour of periodically forcing OnMessageReceived, I decided to then leave it alone (ie, not causing any events to be triggered) for 10 minutes. After 10 minutes transpired, I then re-attempted to trigger OnMessageReceived, but it did not work (and not receiving any errors).

Does this help in narrowing down the problem with Azure?
